modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/InlineUtils.java

```diff
@@ -0,0 +1,186 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.storage.pagememory.index;
+
+import static org.apache.ignite.internal.binarytuple.BinaryTupleCommon.valueSizeToEntrySize;
+import static org.apache.ignite.internal.pagememory.tree.io.BplusInnerIo.CHILD_LINK_SIZE;
+import static org.apache.ignite.internal.util.Constants.KiB;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.List;
+import org.apache.ignite.internal.binarytuple.BinaryTupleCommon;
+import org.apache.ignite.internal.pagememory.freelist.FreeList;
+import org.apache.ignite.internal.pagememory.tree.BplusTree;
+import org.apache.ignite.internal.pagememory.tree.io.BplusInnerIo;
+import org.apache.ignite.internal.pagememory.tree.io.BplusLeafIo;
+import org.apache.ignite.internal.schema.BinaryTuple;
+import org.apache.ignite.internal.schema.NativeType;
+import org.apache.ignite.internal.schema.NativeTypeSpec;
+import org.apache.ignite.internal.schema.VarlenNativeType;
+import org.apache.ignite.internal.storage.index.IndexDescriptor;
+import org.apache.ignite.internal.storage.index.IndexDescriptor.ColumnDescriptor;
+
+/**
+ * Helper class for index inlining.
+ *
+ * <p>Index inlining is an optimization that allows index rows (or prefix) to be compared in a {@link BplusTree} without loading them from a
+ * {@link FreeList}.
+ */
+public class InlineUtils {
+    /** Maximum inline size for a {@link BinaryTuple}, in bytes. */
+    public static final int MAX_BINARY_TUPLE_INLINE_SIZE = 2 * KiB;
+
+    /** Heuristic maximum inline size of a variable length column in bytes. */
+    static final int MAX_VARLEN_INLINE_SIZE = 64;
+
+    /** Heuristic maximum size of an entry in the {@link BinaryTuple} offset table, in bytes. */
+    static final int MAX_BINARY_TUPLE_OFFSET_TABLE_ENTRY_SIZE = 2;
+
+    /**
+     * Minimum number of items in a {@link BplusInnerIo} so that the search in the B+tree does not lose much in performance due to its rapid
+     * growth.
+     */
+    static final int MIN_INNER_PAGE_ITEM_COUNT = 2;
+
+    /** Heuristic maximum inline size for {@link BigDecimal} and {@link BigInteger} column in bytes. */
+    static final int BIG_NUMBER_INLINE_SIZE = 4;
+
+    /**
+     * Calculates inline size for column.
+     *
+     * @param nativeType Column type.
+     * @return Inline size in bytes.
+     */
+    static int inlineSize(NativeType nativeType) {
+        NativeTypeSpec spec = nativeType.spec();
+
+        if (spec.fixedLength()) {
+            return nativeType.sizeInBytes();
```

Comment:
Isn't it possible to define a fixed-length type with huge length? Currently, we have bitmask, can we specify that it has length of 1000 bytes? Anyway, should we add a check that the size of a fixed-length type is not too big for inlining?

Reply:
In theory, the user can do this, but we are limited to the `MAX_BINARY_TUPLE_INLINE_SIZE = 2 * KiB` anyway. 
I can add an assertion that there was no overflow after summing the inline for all columns.
Wdyt?

Reply:
No, I think, `Max.min()` is enough